ykcompile/src/lib.rs

```diff
@@ -0,0 +1,175 @@
+#![feature(proc_macro_hygiene)]
+#![feature(test)]
+
+#[macro_use]
+extern crate dynasm;
+extern crate dynasmrt;
+extern crate test;
+
+use std::collections::HashMap;
+use std::mem;
+
+use yktrace::tir::{TirTrace, TirOp, Statement, Rvalue, Operand, Constant, ConstantInt,
+UnsignedInt};
+
+use dynasmrt::DynasmApi;
+
+/// A compiled SIRTrace.
+pub struct CompiledTrace {
+    /// A compiled trace.
+    mc: dynasmrt::ExecutableBuffer
+}
+
+impl CompiledTrace {
+    pub fn execute(&self) -> u64 {
+        // For now a compiled trace always returns whatever has been left in register RAX. We also
+        // assume for now that this will be a u64.
+        let func: fn() -> u64 = unsafe {
+            mem::transmute(self.mc.ptr(dynasmrt::AssemblyOffset(0)))
+        };
+        func()
+    }
+}
+
+/// The TraceCompiler takes a SIRTrace and compiles it to machine code. Returns a CompiledTrace.
+pub struct TraceCompiler {
+    asm: dynasmrt::x64::Assembler,
+    available_regs: Vec<u8>,
```

Comment:
We could use a bit-field for this to save space?

Reply:
We could. But then we also need a mapper that maps the bits to the correct registers, since currently we are using registers 1, 2, 8, 9, 10, 11.

But all of this is going to change anyways once we have a proper register allocator, so I don't think it's worth optimising things already.

Reply:
Very good point. Let's keep it simple.

Reply:
https://github.com/bnjbvr/regalloc.rs is a crate that is currently being made to be a new register allocator for Cranelift. It has both linear-scan and a backtracking support.

Reply:
That's definitely worth looking into then! Thanks

Reply:
Thanks, this certainly will be useful in the future. One thing I am wondering though is, since we only need to allocate linear traces (so there's no real need for liveness analysis) what advantages would a proper register allocater give us?

For now our rough plan was to use `StorageLive/StorageDead` information from the MIR to allocate/deallocate registers.

Reply:
A proper register allocator tries to optimize when to spill what register in case of register pressure.